### PR TITLE
WebSocket Channel

### PR DESCRIPTION
**Pull request by adamdriscoll**
Implements WebSocket support using OWIN.

This PR isn't quite done but wanted to started a discussion.

Questions:
- WebSocket unit testing is tricky. I can self-host OWIN and test that way but it's not quite a unit test then. 
- Should this be a separate library or is it cool in the main protocol library? It does introduce OWIN dependencies.

### CODE REVIEW

File: PowerShellEditorServices.sln [hunk 1]
```diff
@@ -1,99 +1,99 @@
-﻿
-Microsoft Visual Studio Solution File, Format Version 12.00
```
Comment (daviwil) [inline PowerShellEditorServices.sln]: Weird, is there a line ending conversion happening that makes these files seem like every line changed?  What response do you get from this command: `git config core.autocrlf`

**Comment by adamdriscoll** [comment]
After looking at this PR, I'm going to close it and create one where it's actually readable...

File: PowerShellEditorServices.sln (hunk 1 above)
Comment (adamdriscoll) [inline PowerShellEditorServices.sln]: It returns true. Is that why this merge looks like crap?

Comment (daviwil) [inline PowerShellEditorServices.sln]: Mine returns `true` also so I'm not sure why this would be happening.  Strange!

Comment (adamdriscoll) [inline PowerShellEditorServices.sln]: I'm going to just kill my fork and copy in my changes. It's messy anyways.

**Comment by daviwil** [comment]
Regarding your question about which library this should be in:  I think this might be a good time to establish a new channel library pattern by creating a new PowerShellEditorServices.Channel.WebSocket project under the `src` folder.  I think I'll keep the stdio channel in-box but all other channels can be extensions to the existing design.  Is it possible that the code that's in your Host.IIS library could be put there?

For unit testing, I'm totally fine with self-hosting the socket with OWIN and setting up a local client/server in the unit test code.  I'm doing that pretty extensively with the stdio implementation so I don't see a big difference (so long as we can catch the inevitable configuration problems that self-hosting OWIN on a CI server will cause).

File: PowerShellEditorServices.sln (hunk 1 above)
Comment (daviwil) [inline PowerShellEditorServices.sln]: Check out this StackOverflow answer: http://stackoverflow.com/a/20653073

Maybe core.autocrlf needs to be set to 'input' ?  I've never had to use that before, though

Comment (adamdriscoll) [inline PowerShellEditorServices.sln]: Thanks. I'll futz with it. I have a nice clean fork now. :)

**Comment by adamdriscoll** [comment]
Channel library convention sounds good to me. I'll make that change. As for the host IIS project, I think I'll just get rid of that. It's really just a standard OWIN ASP.NET host. It would make a good example but not a part of the SDK. 

Unit testing: Perfect. I'll spend some time on that after I move things around.

**Comment by daviwil** [comment]
My ultimate goal is to have one set of LanguageServer and DebugAdapter unit tests that get used commonly across all channel and interface implementations, so don't worry about writing a bunch of unit tests for the new channel other than just making sure that the client/server can send/receive messages.  Once I get back to the ILanguageService/IDebugService interface stuff this will all light up pretty fast :)

**Comment by adamdriscoll** [comment]
Good to know. I was thinking of inheriting from them someone how for my tests but if you are working on it already I'll let it be. :D

**Comment by Jaykul** [comment]
By the way @adamdriscoll, this is GitHub -- around here it's perfectly acceptable to rebase and `push -f` your pull requests to clean stuff up ;-)

**Comment by daviwil** [comment]
That's basically what I was going to do, make an abstract base that accepts a channel as constructor parameter.  I've got a working version of it in an unfinished branch.

**Comment by daviwil** [comment]
Haha, yeah, that's what I do, `push -f` about 20 times before I'm happy with it.

**Pull request closed without merging**